src/test/java/lambda/part3/exercise/Mapping.java

```diff
@@ -111,61 +120,210 @@ public void mapping() {
-            return flatMap(listFunction);
-        }
-
-        // (R -> R2) -> (R -> [R2])
-        private <R2> Function<R, List<R2>> rR2TorListR2(Function<R, R2> f) {
-            throw new UnsupportedOperationException();
+            return new LazyFlatMapHelper<>(list, traversable.map(f));
         }
 
         // TODO *
         public <R2> LazyFlatMapHelper<T, R2> flatMap(Function<R, List<R2>> f) {
-            throw new UnsupportedOperationException();
+            return new LazyFlatMapHelper<>(list, traversable.flatMap(f));
+        }
+    }
+
+    interface Traversable<T> {
+        void forEach(Consumer<T> c);
+
+        default <R> Traversable<R> map(Function<T, R> f) {
+            return (c) -> this.forEach(t -> c.accept(f.apply(t)));
+        }
+
+        default <R> Traversable<R> flatMap(Function<T, List<R>> f) {
+            return (c) -> this.forEach(t -> f.apply(t).forEach(c));
+        }
+
+        default Traversable<T> filter(Predicate<T> p) {
+            return (c) -> this.forEach(t -> {
+                if (p.test(t)) c.accept(t);
+            });
+        }
+
+        default List<T> force() {
+            List<T> result = new ArrayList<>();
+            forEach(result::add);
+            return result;
+        }
+
+        static <T> Traversable<T> from(List<T> l) {
+            return l::forEach;
+        }
+    }
+
+    interface ReachIterable<T> {
+        boolean forNext(Consumer<T> c);
+
+        static <T> ReachIterable<T> from(List<T> l) {
+            final Iterator iterator = l.iterator();
+            return c -> {
+                if (iterator.hasNext()) {
+                    c.accept((T) iterator.next());
+                    return true;
+                } else {
+                    return false;
+                }
+            };
+        }
+
+        default List<T> force() {
+            List<T> result = new ArrayList<T>();
+            while (forNext(result::add)) ;
+            return result;
+        }
+
+        // filter
+        default ReachIterable<T> filter(Predicate<T> p) {
+            final ReachIterable<T> prev = this;
+            return c -> {
+                final boolean[] found = {false};
+                boolean hasNext = true;
+                while (!found[0] && hasNext) {
+                    hasNext = prev.forNext(t -> {
+                        if (p.test(t)) {
+                            c.accept(t);
+                            found[0] = true;
+                        }
+                    });
+                }
+                return hasNext;
+            };
+        }
+
+        // map
+        default <R> ReachIterable<R> map(Function<T, R> f) {
+            final ReachIterable<T> prev = this;
+            return c -> prev.forNext(t -> {
+                c.accept(f.apply(t));
+            });
+        }
+
+        // flatMap
+        default <R> ReachIterable<R> flatMap(Function<T, List<R>> f) {

```

Comment:
Incorrect. Add test with at least 2 elements in initial list and `flatMap(Collections::singletonList)`.